zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/repo/GitNotebookRepo.java

```diff
@@ -96,9 +100,37 @@ public Revision checkpoint(String pattern, String commitMessage) {
   }
 
   @Override
-  public Note get(String noteId, Revision rev) throws IOException {
-    //TODO(bzz): something like 'git checkout rev', that will not change-the-world though
-    return super.get(noteId);
+  public synchronized Note get(String noteId, Revision rev) throws IOException {
```

Comment:
actually there're comments in code with similar explanation, but i also added javadoc on top with step-by-step explanation to be more clear. regarding the conflict case, i'm not exactly sure which scenario you mean, but there're shouldn't be any conflicts since changes are linear in time. maybe you could check the javadoc and describe the scenario you mean once again.